### PR TITLE
feat(ec2): filtering selected subnets by availability zone

### PR DESCRIPTION
**Pull request by flemjame-at-amazon**
## Commit Message
feat(ec2): filtering selected subnets by availability zone

This allows a user to specify that their subnet selection should only return subnets that are in the specified availability zones. I added this to provide a native solution to the problem of VPC endpoint services not existing in all AZs. The way CDK currently creates VPC endpoints, it places one in every AZ, which for some services results in this CloudFormation error on deployment:

> The VPC endpoint service <-service name-> does not support the availability zone of the subnet

CDK cannot currently know the subnets in advance, so I added functionality to SubnetSelection to allow for choosing the availability zones, like so:

```ts
new InterfaceVpcEndpoint(stack, "VPC Endpoint", {
  vpc,
  service: new InterfaceVpcEndpointService("com.amazonaws.vpce.us-east-1.vpce-svc-uuddlrlrbastrtsvc", 443),
  subnets: {
    subnetType: SubnetType.ISOLATED,
    availabilityZones: ["us-east-1a", "us-east-1c"]
  }
```

This provides the user a way to solve this problem, because they can get the endpoint service availability zones from the AWS console.

## End Commit Message

This change also includes an update to the README explaining how to use SubnetSelection

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/@aws-cdk/aws-ec2/README.md [hunk 1]
```diff
@@ -88,6 +88,53 @@ itself to 2 Availability Zones.
...
+    subnets: [subnet1, subnet2]
+  }
+});
+```
+
+Which subnets are selected is evaluated as follows:
+
+* Specifying particular subnets overrides any other settings, such as type or 
+availability zone.
+* If no subnets are specified, but a subnetGroupName is, the
```
Comment (rix0rrr) [inline packages/@aws-cdk/aws-ec2/README.md:126]: Thanks for documenting the rules! 

This is in fact not how it's intended. `subnetType` and `subnetGroupName` are supposed to live at the same level of precedence, where `subnetGroupName` is necessary if there are multiple *groups* with the same *type*. (subnetGroup implies subnetType). 

I think `availabilityZones` should work as an additional restrictor applied at the end, similar to when `onePerAz` is applied. So the precedence should work like this:

* Specific subnets, if given. 
* A set of subnet (groups), implied by either subnetType, subnetGroupName, or the absence of both.
  * A subset of those subnets (hah), determined by both `availabilityZones` and/or `onePerAz`.

Comment (flemjame-at-amazon) [reply packages/@aws-cdk/aws-ec2/README.md:126]: I see what you mean. I wrote the readme as a direct reflection of the code that does the selection ([link](https://github.com/aws/aws-cdk/blob/87b16350ed2ccea2cda03e3515cd8797c8047186/packages/%40aws-cdk/aws-ec2/lib/vpc.ts#L386-L405)) but I like yours better.

Comment (flemjame-at-amazon) [reply packages/@aws-cdk/aws-ec2/README.md]: I've updated the readme accordingly, and also modified the selection code such that it reflects that selection is primarily by group name or type, then with additional filtering by availability zone and one per AZ if specified.

**Comment by rix0rrr** [comment]
Thanks! I'm going to rewrite the README slightly I think, but this looks great.

**Pull request merged**